src/backend/stream.rs

```diff
@@ -0,0 +1,821 @@
+                                                           self.user_ptr,
+                                                           read_ptr as *const _ as *mut _,
+                                                           buffer,
+                                                           (size / frame_size) as i64) };
+            if got < 0 {
+                pa_stream_cancel_write(s);
+                self.shutdown = true;
+                return;
+            }
+            // If more iterations move offset of read buffer
+            if !input_data.is_null() {
+                let in_frame_size = pa_frame_size(&self.input_sample_spec);
+                read_offset += (size / frame_size) * in_frame_size;
+            }
+
+            if self.volume != PULSE_NO_GAIN {
+                let samples = (self.output_sample_spec.channels as usize * size / frame_size) as isize;
+
+                if self.output_sample_spec.format == PA_SAMPLE_S16BE ||
+                    self.output_sample_spec.format == PA_SAMPLE_S16LE
+                {
+                    let b = buffer as *mut i16;
+                    for i in 0..samples {
+                        unsafe { *b.offset(i) *= self.volume as i16 };
+                    }
+                } else {
+                    let b = buffer as *mut f32;
+                    for i in 0..samples {
+                        unsafe { *b.offset(i) *= self.volume };
+                    }
+                }
+            }
+
+            r = pa_stream_write(s, buffer, got as usize * frame_size, None, 0, PA_SEEK_RELATIVE);
+            debug_assert!(r == 0);
+
+            if (got as usize) < size / frame_size {
+                let mut latency: pa_usec_t = 0;
+                let rr: i32 = pa_stream_get_latency(s, &mut latency, ptr::null_mut());
+                if rr == -(PA_ERR_NODATA as i32) {
+                    /* this needs a better guess. */
+                    latency = 100 * PA_USEC_PER_MSEC;
+                }
+                debug_assert!(r == 0 || r == -(PA_ERR_NODATA as i32));
+                /* pa_stream_drain is useless, see PA bug# 866. this is a workaround. */
+                /* arbitrary safety margin: double the current latency. */
+                debug_assert!(self.drain_timer.is_null());
+                self.drain_timer =
+                    pa_context_rttime_new(self.context.context,
+                                          pa_rtclock_now() + 2 * latency,
+                                          Some(stream_drain_callback),
+                                          self as *const _ as *mut _);
+                self.shutdown = true;
+                return;
+            }
+
+            towrite -= size;
+        }
+
+        debug_assert!(towrite == 0);
+    }
+}
+
+unsafe extern fn stream_success_callback(_s: *mut pa_stream, _success: i32, u: *mut c_void)
+{
+  let stm = &*(u as *mut Stream);
```

Comment:
Indentation is much shallower for this and some of the following code vs the elsewhere